### ISSUE TITLE
Password security flaw

### ISSUE THREAD

**Issue by SkyKwas** [keepassxreboot/keepassxc#7658]
Before I begin, I want to start off by saying this is my first time reporting an issue on GitHub. Please forgive me if I messed something up. Thank you! :)

## Background

I have recently been taking steps to improve my privacy, security, anonymity, etc. One of these steps is switching from Google Chrome's built-in password manager to KeePassXC (KPXC) because I want to have full control of my passwords; I even bought two OnlyKeys to go with KPXC. I managed to transfer all my passwords from Chrome to KPXC without much trouble and got it working with LibreWolf.

However, I later noticed something that frustrated me for a while before realizing that this is a problem: there is no way to protect the plaintext of passwords when a database is open. This, to me, is a CRITICAL security flaw in KPXC; this is something that even Chrome has!

For context, when you navigate to your password list in Chrome (Settings > Autofill > Passwords), it shows a list of all saved passwords with columns: Site, Username, Password. If you want to view/copy/edit a password, a pop-up appears asking for the computer's password. This prevents someone from... well... viewing, copying, and editing your passwords if they had access to your computer. If you were to let friends/family borrow your computer, they will not be able to get your passwords (as long as they don't know the password to the computer). This can also minimize threats from spyware, and reverse shells if that were to happen.

In KPXC, when a database is open (which it has to be in order to use it), it's just like having a plaintext document containing everything on your computer. Anyone can view/copy/edit anything in the open database without needing to type in a password or tap a hardware key like OnlyKey.

## Summary

I believe this is something that should be fixed ASAP!
So, I would like to request the following:
1. Add security settings in [database settings] such as:

- 1a. Add the option to require a hardware key press and/or keyfile and/or password (authentication) to view/copy/edit Groups, Titles, Usernames, Passwords, URLs, etc. in open databases (settings for each of these would be nice). Currently, Windows asks for the computer password when opening a database for the first time and future times depending on certain settings in KPXC. For me, it's database password + OnlyKey + Windows/Microsoft password. On Linux Mint, it's just database password + OnlyKey (OnlyKey app needed). Ideally, I would like to have the option to require a hardware key press AND computer password to view/copy/edit databases.

- 1b. Add the option to completely hide Groups, Titles, Usernames, Passwords, URLs, etc. in open databases (settings for each of these would be nice), not just putting things in the form  "******************", but actually preventing anyone/anything from even seeing the lengths.

- 1c. For (1a), add the option to require authentication: per entry AND/OR every XXX minutes AND/OR every XXX views/copies/edits. 

- 1d. Add a [lock database views and changes] button in the top bar next to the [lock databases] button to instantly lock views/copies/edits privileges, requiring authentication to unlock the open database for future views/copies/edits. The format of this button is the same as the current [lock database] button where clicking the lock icon locks the current database, and has a drop-down menu to lock all databases (from views and changes NOT locking the database itself).

- 1e. For (1c), add the option to require authentication for specific entries.

2. Require authentication to view and edit settings in [database settings].

3. Add the option to require authentication when adding an entry via the KPXC-browser addon.

## Examples
If these changes were to happen, what would be different? For the most part, not much. Opening the database is the same as it was before. The KPXC-browser addon works the same and doesn't need extra authentication because you can't view/copy/edit passwords it puts in the browser.

The only time you will need to put in extra authentication is when you need to view/change the contents the database directly. The level of convenience depends on what options there are and how many there are.

## Final Words
I've only been using KPXC for a few days, but can already see its potential. I'm excited to use it, but this flaw quickly caught my eye and I felt the need to address it.

To sum up:
Currently in KPXC, an open database file is the same thing as having a plaintext file containing all your sensitive information (very bad).

**Comment by droidmonkey** [comment]
This is not a flaw, this is by design and required to provide the functions of the application. We secure the applications memory from other programs. We also prevent screenshots of the application interface. You can enable auto lock after timeout to reprotect your database after a period of non use.

Thank you for the report but this is not a bug or a security issue.

**Comment by michaelk83** [comment]
> You can enable auto lock after timeout to protect your database after a period of non use.

Combine with QuickUnlock on Windows/MacOS (not yet available on Linux). You can also set it to lock when the session is locked, and assign a global shortcut to quickly lock your session when you leave the PC.

**Comment by SkyKwas** [comment]
Oh, okay. Thank you for letting me know about these security features. I tried to record KPXC with OBS and it gave me a black screen. Cool.

As for the auto-lock, KPXC maxes out at 9999 seconds, I was wanting to auto-lock after 24 hours so I don't have to login again for the rest of the day.

So I guess just don't share your computer when your databases are open. I still feel like it would be a nice feature to censor information and still use an open database.

**Comment by michaelk83** [comment]
You were asking to add authentication for various operations and hide the information otherwise. This is the same as locking the database with QuickUnlock.

**Comment by SkyKwas** [comment]
Oh, no no, I don't want to lock the database, I just want to hide the information in it while still being able to use it.

I'm trying to think of the best way to explain it...

Okay, so right now, when a database is open, you can view/copy*/edit all of the information in it and use things like the KPXC-browser addon. When a database is locked, you can't do any of that stuff.

*When I say copy in this case, I mean the ability to get the plaintext via clipboard.

What I would like is the addition of an option to require authentication a second time to a database that is already open to do things like view passwords in plaintext on the screen and change them. I know that I just learned KPXC protects against screen-capture software, but this does nothing to protect against someone who has physical access to the computer with a database already open.

It would be like this:

Databases would have three states instead two:
1) Locked database
2) Open/Usable database
3) Open/Editable database

A locked database would be the same as before: can't view/copy/edit any information. Authentication is required to go from a locked database to an open/usable database (1 -> 2).

An open/usable database would be a limited form of the open/editable database. In this state, information like passwords will be unable to be viewed in plaintext and can't be copied to the clipboard nor will they be able to edit. However, they are still usable. You will still be able to log into your accounts using the KPXC-browser addon, or using the double-click auto-fill option in KPXC without exposing the plaintext to the clipboard. In other words, this database is in a USABLE state, you can use the information inside it, but you don't know what the information is. Authentication is required to go from an open/usable database to an open/editable database (2 -> 3).

An open/editable database would be how open databases are now: you can view/copy/edit all of the information inside it (you have full control of the database).

So, if all you need to do is use the database to log into your accounts, you only need to add authentication once (just like how it is right now). But if you want to view/copy/edit the database in any way, you would need to add authentication again.

Did I explain it better or did I leave you more confused?

**Comment by michaelk83** [comment]
This has been discussed before. You have two options:
1. Leave the DB completely open indefinitely.
2. Open -> use -> (auto) lock until it's needed again -> repeat. Most of the time the DB is open, it's not actually being used, so better to lock it. QuickUnlock makes this less painful, since you don't need to enter the full passphrase every time, and auto-lock takes care of the locking.

**Comment by phoerious** [comment]
This is not how KeePassXC works. A password manager is not something you let other people browse in if you don't fully trust them. Either you lock your database or you leave it open, but there is no in-between.

**Comment by SkyKwas** [comment]
It's not that I let other people use my passwords, I just wanted to add an extra layer of protection in case the system becomes compromised. That's all.

**Comment by phoerious** [comment]
There is no protection if your system is compromised. We already do what's possible to mitigate the impact, but all in all, your data does not belong to you anymore if that happens.

**Comment by SkyKwas** [comment]
Okay, yeah... I see your point. It would be redundant to hide the information if you could still log into your services. The reason why I brought this issue up in the first place was because Chrome does this and I thought "oh, that's smart" and was surprised KPXC doesn't. But now I understand your point better.

I still think it would be a nice feature though, but I understand why this isn't as much as a security concern as I originally thought.

**Comment by phoerious** [comment]
Chrome does it, because its primary purpose is something else (browsing websites, mostly). The equivalent of Chrome asking for your password is unlocking your KeePassXC database.
